src/main/java/hudson/plugins/ansicolor/ColorConsoleAnnotator.java

```diff
@@ -66,29 +66,33 @@
             CountingOutputStream outgoing = new CountingOutputStream(new NullOutputStream());
             class EmitterImpl implements AnsiAttributeElement.Emitter {
                 CountingOutputStream incoming;
-                int adjustment;
+                int multiByteCharAdjustment;
+                int hiddenCharAdjustment;
                 int lastPoint = -1; // multiple HTML tags may be emitted for one control sequence
                 @Override
                 public void emitHtml(String html) {
-                    LOGGER.finest("emitting " + html + " @" + incoming.getCount());
-                    text.addMarkup(incoming.getCount(), html);
-                    if (incoming.getCount() != lastPoint) {
-                        lastPoint = incoming.getCount();
-                        int hide = incoming.getCount() - outgoing.getCount() - adjustment;
-                        LOGGER.finest("hiding " + hide + " @" + (outgoing.getCount() + adjustment));
-                        text.addMarkup(outgoing.getCount() + adjustment, outgoing.getCount() + adjustment + hide, "<!--", "-->");
-                        adjustment += hide;
+                    int inCount = incoming.getCount() - multiByteCharAdjustment;
+                    int outCount = outgoing.getCount() - multiByteCharAdjustment + hiddenCharAdjustment;
+                    LOGGER.log(Level.FINEST, "emitting {0} @{1}/{2}", new Object[] { html, inCount, s.length() });
+                    text.addMarkup(inCount, html);
+                    if (inCount != lastPoint) {
+                        lastPoint = inCount;
+                        int hide = inCount - outCount;
+                        LOGGER.log(Level.FINEST, "hiding {0} @{1}", new Object[] { hide, outCount });
+                        text.addMarkup(outCount, outCount + hide, "<!--", "-->");
+                        hiddenCharAdjustment += hide;
                     }
                 }
             }
             EmitterImpl emitter = new EmitterImpl();
             CountingOutputStream incoming = new CountingOutputStream(new AnsiHtmlOutputStream(outgoing, colorMap, emitter));
             emitter.incoming = incoming;
             try {
-                byte[] data = s.getBytes(charset);
-                for (int i = 0; i < data.length; i++) {
-                    // Do not use write(byte[]) as offsets in incoming would not be accurate.
-                    incoming.write(data[i]);
+                for (int i = 0; i < s.length(); i++) {
+                    // We write one UTF-16 code unit at a time so that our counting methods can handle multibyte characters correctly.
+                    byte[] chars = String.valueOf(s.charAt(i)).getBytes(charset);
```

Comment:
Also, maybe we need iterate through the Unicode code points rather than the UTF-16 code units in case any characters are not in the BMP? I tried adding emoji to the test case, but it looks like the emoji is converted into the replacement character for unmappable characters somewhere along the line before `annotate` is called (not sure where), so I don't think it's possible to handle it correctly here (although maybe I was doing something wrong in my test).

Reply:
I suspect this entire (`src/main/`) patch could be simplified to

```diff
-byte[] data = s.getBytes(charset);
+byte[] data = s.getBytes("US-ASCII");
```

plus deleting `charset` and the code that sets it. After all, we do not really care what any of the >U+007F characters in the log are for this purpose—we are throwing out the output and just looking for ESC and stuff, so if we send some `(byte) '?'` through it will not matter.

(Alternately, do not bother getting a `byte[]` at all, just loop over `char`s in `s` and write one `byte` for each—either the same, if ASCII, or something arbitrary like `'?'` otherwise.)

Reply:
Ok, it seemed easiest to just switch to using US-ASCII so I did that and added an explanatory comment.